org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/XMLWorkspaceService.java

```diff
@@ -12,32 +12,58 @@
  */
 package org.eclipse.lemminx;
 
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
 
+import org.eclipse.lemminx.services.IXMLCommandService;
 import org.eclipse.lsp4j.DidChangeConfigurationParams;
 import org.eclipse.lsp4j.DidChangeWatchedFilesParams;
+import org.eclipse.lsp4j.ExecuteCommandParams;
 import org.eclipse.lsp4j.FileEvent;
 import org.eclipse.lsp4j.SymbolInformation;
 import org.eclipse.lsp4j.WorkspaceSymbolParams;
 import org.eclipse.lsp4j.services.WorkspaceService;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 /**
  * XML workspace service.
  *
  */
-public class XMLWorkspaceService implements WorkspaceService {
+public class XMLWorkspaceService implements WorkspaceService, IXMLCommandService {
+	
+	private static final String WORKSPACE_EXECUTE_COMMAND = "workspace/executeCommand";
 
 	private final XMLLanguageServer xmlLanguageServer;
-
+	
+	private final Map<String, Function<ExecuteCommandParams, CompletableFuture<Object>>> commands;

```

Comment:
I think command handler should receive 2 parameters:

 * ExecuteCommandParams like you did
 * the CancelChecker

Reply:
IMHO I think we should provide an new interface `IDelegateCommandHandler` like https://github.com/eclipse/eclipse.jdt.ls/blob/cf6358ba4c166fa7d4c45aba5b7b22ea08516524/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/IDelegateCommandHandler.java

```java
public interface IDelegateCommandHandler {

    CompletableFuture<Object> executeCommand(ExecuteCommandParams params, ICancelChecker cancelChecker) throws Exception;
}
```

instead of using Function which is very generic. No?

Reply:
Not sure about this one... JDT LS uses progress monitor for command execution because: https://github.com/eclipse/eclipse.jdt.ls/blob/e69383c16c10ed6e3726ae8d358e1431147af313/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java#L978
In other words they have progress support message (LSP extension). There is progress message support in the LSP spec but in the LSP4J yet. I don't see how `ICancelChecker` would be used at the moment. I could introduce `IDelegateCommandHandler` instead of the function but without the `ICancelChecker`. However, i only see the handler being used as a lambda and therefore thought function is ok. Anyway, I don't mind `IDelegateCommandHandler` but i'd go without the `ICancelChecker`

Reply:
>  I don't see how ICancelChecker would be used at the moment. 

To check if client has cancel the command process.  For progress feature, we will have those information with ExecuteCommandParams.

Reply:
Is the cancel coming in as `$/cancelRequest`?

Reply:
It was my understand. See https://github.com/eclipse/lsp4j/tree/master/documentation#cancellation-support

Reply:
Thanks!-) The link clarified it for me :-) So this addressed as well now